File: xeddy-ios/xeddy/Models/Restaurant.swift
//
//  Restaurant.swift
//  xeddy
//
//  Created by David Ledbetter on 9/10/24.
//

// need coding keys to ensure the (Codable, Identifiable) types remain valid for swift
// syntax checks.

import Foundation

struct MenuItem: Identifiable, Codable {
    let id: String
    let name: String
    let description: String
    let priceMoney: Int // needs to updated to support priceMoney handler.
    var imageURL: URL?

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case description
        case priceMoney = "price_money"
        case imageURL = "image_url"
    }

}

struct LineItem: Codable {
    let id: String
    let quantity: String
    let base_price_money: Money
}

struct Money: Codable {
    let amount: Int // Amount in cents (e.g., $9.95 should be 995)
    let currency: String // "USD"
}

extension MenuItem {
    func toLineItem(quantity: Int) -> LineItem {
        return LineItem(
            name: self.name,
            quantity: String(quantity),
            base_price_money: Money(amount: self.price_money, currency: "USD")
        )
    }
}

struct TimeObject: Codable {
    let open: String
    let close: String

    enum CodingKeys: String, CodingKey {
        case open = "Open"
        case close = "Close"
    }
}

struct Item: Identifiable, Codable {
    let id: Int // ID as an integer
    let name: String
    let description: String
    let price: Double // Double to handle decimal prices

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case name = "Name"
        case description = "Description"
        case price = "Price"
    }
}

struct Deal: Codable, Identifiable {
    let id: Int64
    let description: String
    let validUntil: Date

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case description = "Description"
        case validUntil = "ValidUntil"
    }
}

struct Review: Codable, Identifiable {
    let id: Int64
    let rating: Int
    let text: String

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case rating = "Rating"
        case text = "Text"
    }
}

struct Restaurant: Codable, Identifiable {
    let id: Int64
    let name: String
    let category: String
    let imageURL: String
    let logoURL: String
    let location: String
    let descriptionLong: String
    let descriptionShort: String
    let email: String
    let phone: String
    let merchantID: String
    let oddDates: [Date]?
    let hours: [String: TimeObject]?
    let closed: Bool
    let menu: [Item]?
    let rewards: [Int64]?
    let deals: [Deal]?
    let orderSystems: [String]?
    let reviews: [Review]?

    enum CodingKeys: String, CodingKey {
        case id = "ID"
        case name = "Name"
        case category = "Category"
        case imageURL = "ImageURL"
        case logoURL = "LogoURL"
        case location = "Location"
        case descriptionLong = "DescriptionLong"
        case descriptionShort = "DescriptionShort"
        case email = "Email"
        case phone = "Phone"
        case merchantID = "MerchantID"
        case oddDates = "OddDates"
        case hours = "Hours"
        case closed = "Closed"
        case menu = "Menu"
        case rewards = "Rewards"
        case deals = "Deals"
        case orderSystems = "OrderSystems"
        case reviews = "Reviews"
    }
}

File: xeddy-backend/internal/models/restaurant.go
// const (
//     first = iota  // first is 0
//     second        // second is 1
//     third         // third is 2
// )

// func main() {
//     fmt.Println(first, second, third)  // Outputs: 0 1 2
// }

// models/restaurant.go

package models

import "time"

// Use string instead of custom WeekDay type for Firestore compatibility
type TimeObject struct {
	Open  string
	Close string
}

type Item struct {
	ID          int64
	Name        string
	Description string
	Price       float64
}

type Deal struct {
	ID          int64
	Description string
	ValidUntil  time.Time
}

type Review struct {
	ID     int64
	Rating int
	Text   string
}

type Restaurant struct {
	ID               int64
	Name             string
	Category         string
	ImageURL         string
	LogoURL          string
	Location         string
	DescriptionLong  string
	DescriptionShort string
	Email            string
	Phone            string
	OddDates         []time.Time
	Hours            map[string]TimeObject // Use string instead of WeekDay
	Closed           bool
	Menu             []Item
	Rewards          []int64
	Deals            []Deal
	OrderSystems     []string
	Reviews          []Review
}

func ExampleRestaurant() Restaurant {
    return Restaurant{
        ID:               1,
        Name:             "Gusto Italian Grill",
        Category:         "Italian",
        ImageURL:         "https://example.com/image.jpg",
        LogoURL:          "https://example.com/logo.jpg",
        Location:         "1234 Culinary Blvd, Taste City, TC 56789",
        DescriptionLong:  "Gusto Italian Grill offers a rich taste of Italy with a diverse menu featuring traditional dishes, handpicked wines, and a cozy ambiance perfect for any occasion.",
        DescriptionShort: "Traditional Italian cuisine in the heart of Taste City.",
        Email:            "contact@gustoitalian.com",
        Phone:            "555-1234",
        OddDates:         []time.Time{time.Now(), time.Now().AddDate(0, 0, 1)},
        Hours: map[string]TimeObject{
            "Monday":  {Open: "10:00:00", Close: "22:00:00"},
            "Tuesday": {Open: "10:00:00", Close: "22:00:00"},
        },
        Closed: false,
        Menu: []Item{
            {ID: 1, Name: "Spaghetti Carbonara", Description: "Creamy pasta with pancetta, egg, and Parmesan", Price: 15.99},
            {ID: 2, Name: "Margherita Pizza", Description: "Classic pizza with fresh tomatoes, mozzarella, and basil", Price: 12.50},
        },
        Rewards: []int64{100, 200},
        Deals: []Deal{
            {ID: 1, Description: "20% off on Wednesdays", ValidUntil: time.Now().AddDate(0, 1, 0)},
        },
        OrderSystems: []string{"Toast", "Square"},
        Reviews: []Review{
            {ID: 1, Rating: 5, Text: "Outstanding service and food quality!"},
        },
    }
}